src/views/memory-webview-view.tsx

```diff
@@ -62,41 +65,47 @@ class App extends React.Component<{}, MemoryState> {
     }
 
     public render(): React.ReactNode {
-        const { memory } = this.state;
-        return (
-            <div>
-                <MemoryTable memory={memory}>
-                </MemoryTable>
-            </div>
-        );
+        return <MemoryWidget
+            memory={this.state.memory}
+            memoryReference={this.state.memoryReference}
+            offset={this.state.offset}
+            count={this.state.count}
+            updateMemoryArguments={this.justSetState}
+            refreshMemory={this.refreshMemory}
+        />;
     }
 
-    protected async setOptions(options: MemoryOptions): Promise<void> {
+    protected justSetState = (newState: Partial<MemoryState>) => this.setState(prevState => ({ ...prevState, ...newState }));
+
+    protected async setOptions(options?: Partial<DebugProtocol.ReadMemoryArguments>): Promise<void> {
         this.messenger.sendRequest(logMessageType, HOST_EXTENSION, JSON.stringify(options));
+        this.setState(prevState => ({ ...prevState, ...options }));
+        return this.fetchMemory(options);
+    }
 
-        const response = await this.messenger.sendRequest(readMemoryType, HOST_EXTENSION, {
-            memoryReference: `${options.startAddress}`,
-            count: options.readLength,
-            offset: options.locationOffset
-        });
+    protected refreshMemory = () => { this.fetchMemory(); };
+
+    protected async fetchMemory(partialOptions?: Partial<DebugProtocol.ReadMemoryArguments>): Promise<void> {
+        const completeOptions = {
+            memoryReference: partialOptions?.memoryReference || this.state.memoryReference,
+            offset: partialOptions?.offset ?? this.state.offset,
+            count: partialOptions?.count ?? this.state.count
+        };
+
+        const response = await this.messenger.sendRequest(readMemoryType, HOST_EXTENSION, completeOptions);
 
         this.setState({
             memory: this.convertMemory(response)
         });
     }
 
-    protected convertMemory(result: MemoryReadResponse): Memory {
-        if (result.address.startsWith('0x')) {
-            // Assume hex
-            const bytes = Uint8Array.from(Buffer.from(result.data, 'hex'));
-            const address = Long.fromString(result.address, true, 16);
-            return { bytes, address };
-        } else {
-            // Assume base64
-            const bytes = Uint8Array.from(Buffer.from(result.data, 'base64'));
-            const address = Long.fromString(result.address, true, 10);
-            return { bytes, address };
-        }
+    protected convertMemory(result: DebugProtocol.ReadMemoryResponse['body']): Memory {
+        if (!result?.data) { throw new Error('No memory provided!'); }
+        const address = result.address.startsWith('0x')
+            ? Long.fromString(result.address, true, 16)
+            : Long.fromString(result.address, true, 10);
+        const bytes = Uint8Array.from(Buffer.from(result.data, 'base64'));
```

Comment:
This change loses the conversion from hex when the address starts with `0x`:

e.g. in that case should it be:

```typescript
const bytes = Uint8Array.from(Buffer.from(result.data, 'hex'));
```

Reply:
The bytes are never supposed to be hex encoded, only the address may be. Cf. https://microsoft.github.io/debug-adapter-protocol/specification#Requests_ReadMemory